dbfopen.c

```diff
@@ -1311,9 +1311,18 @@ static bool DBFWriteAttribute(DBFHandle psDBF, int hEntity, int iField,
     /* -------------------------------------------------------------------- */
     if (pValue == SHPLIB_NULLPTR)
     {
-        memset(pabyRec + psDBF->panFieldOffset[iField],
-               DBFGetNullCharacter(psDBF->pachFieldType[iField]),
-               psDBF->panFieldSize[iField]);
+        if (psDBF->pachFieldType[iField] == 'D')
+        {
+            // NULL Date is independent of panFieldSize
+            pabyRec += psDBF->panFieldOffset[iField];
+            pabyRec = SHPLIB_NULLPTR;
+        }
+        else
+        {
+            memset(pabyRec + psDBF->panFieldOffset[iField],
+                   DBFGetNullCharacter(psDBF->pachFieldType[iField]),
+                   psDBF->panFieldSize[iField]);
+        }
```

Comment:
This change looks completely wrong to me

Reply:
As I said, other fixes are possible to get the failing tests green.

Reply:
I don't understand why it is necessary to change that part of the code. What's wrong with the current code? As far as I understand, it should fill the date field with the '0' character, which is the expected behavior

Reply:
No, `DBFWriteNULLAttribute` writes a string of  "0" repeated `panFieldSize` times, but only "0" and "00000000" is considered NULL Date in `DBFIsAttributeNULL `.

Reply:
> `DBFWriteNULLAttribute` writes a string of "0" repeated `panFieldSize` times

That should be fine. panFieldSize[] should be 8 for a date field

Reply:
> That should be fine. panFieldSize[] should be 8 for a date field

Which is not checked or verified. `DBFWriteNULLAttribute` could be fixed to return false on unexpected `panFieldSize`.

Reply:
wouldn't a better fix to check that the size is 8 for date fields in DBFAddNativeFieldType ?

Reply:
Sounds very reasonable to me. However I am not sure if it breaks existing applications. Should we also allow size = 1 (such that " " and "0" can be successfully read as NULL Date strings?

Reply:
There's indeed a slight breakage possibility, but I've skimmed a bit various DBF documentation, and all document Date fields being 8 byte large, so anyone producing such files would produce unusual/non-conformant files. If we keep that restriction on DBFAddNativeFieldType(), we still allow reading those unusual/non-conformant files

The alternatives would be:
- not care about DBFIsAttributeNULL() not working for date field of size != 8
- or make DBFIsAttributeNULL() tolerant to be set to as many '0' as the field width

Reply:
I went for the second alternative and updated this PR accordingly.